### PR TITLE
Trick to load .config.so file in ANDROID instead of .config

### PR DESCRIPTION
**Pull request by viniciusmarangoni**
Frida gadget looks for a config file with the same name of the lib. In Android, during installation, the files that do not starts with "lib" and do not ends with ".so" are not copied to libs folder. With this workaround the gadget will look to <gadgetname>.config.so instead of <gadgetname>.config

So, if your gadget file name is "libfrida-gadget.so", the gadget will look for "libfrida-gadget.config.so" config file.

(Tested in Android 7.0)

### CODE REVIEW

File: lib/gadget/gadget.vala [hunk 1]
```diff
@@ -494,6 +494,25 @@ namespace Frida.Gadget {
 	private Config load_config (Location location) throws Error {
 		var config_path = derive_config_path_from_file_path (location.path);
 
+		if(FileUtils.test (config_path, FileTest.EXISTS)) {
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: Missing space after `if`.

File: lib/gadget/gadget.vala [hunk 2]
```diff
@@ -494,6 +494,25 @@ namespace Frida.Gadget {
...
+			log_info ("Config file found: %s".printf (config_path));
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: Let's avoid introducing additional logging.

File: lib/gadget/gadget.vala [hunk 3]
```diff
@@ -494,6 +494,25 @@ namespace Frida.Gadget {
...
+		}
+		else {
+			log_warning ("Config file not found: %s".printf (config_path));
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: Let's avoid introducing additional logging.

File: lib/gadget/gadget.vala [hunk 4]
```diff
@@ -494,6 +494,25 @@ namespace Frida.Gadget {
 	private Config load_config (Location location) throws Error {
 		var config_path = derive_config_path_from_file_path (location.path);
 
+		if(FileUtils.test (config_path, FileTest.EXISTS)) {
+			log_info ("Config file found: %s".printf (config_path));
+		}
+		else {
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: Style is off, should be:

```Vala
		} else {
```

File: lib/gadget/gadget.vala [hunk 5]
```diff
@@ -494,6 +494,25 @@ namespace Frida.Gadget {
...
+			log_warning ("Config file not found: %s".printf (config_path));
+
+			#if ANDROID
+				var ext_index = config_path.last_index_of_char ('.');
+				if (ext_index != -1) {
+					config_path = config_path[0:ext_index] + ".config.so";
+				}
+				else{
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: Style (see above).

File: lib/gadget/gadget.vala [hunk 6]
```diff
@@ -494,6 +494,25 @@ namespace Frida.Gadget {
...
+					config_path = config_path + ".config.so";
+				}
+
+				log_info ("Trying to load the config file with the alternative extension: %s".printf (config_path));
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: Let's avoid introducing additional logging.

File: lib/gadget/gadget.vala [hunk 7]
```diff
@@ -494,6 +494,25 @@ namespace Frida.Gadget {
 	private Config load_config (Location location) throws Error {
 		var config_path = derive_config_path_from_file_path (location.path);
 
+		if(FileUtils.test (config_path, FileTest.EXISTS)) {
+			log_info ("Config file found: %s".printf (config_path));
+		}
+		else {
+			log_warning ("Config file not found: %s".printf (config_path));
+
+			#if ANDROID
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: `#if` and `#endif` should not be indented.

File: lib/gadget/gadget.vala [hunk 8]
```diff
@@ -1541,7 +1560,8 @@ namespace Frida.Gadget {
 		else
 			stem = filename;
 
-		return Path.build_filename (dirname, stem + ".config");
+		var configpath = Path.build_filename (dirname, stem + ".config");
+		return configpath;
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: Unrelated change.

File: lib/gadget/gadget.vala (hunk 1 above)
Comment (oleavr) [inline lib/gadget/gadget.vala]: Also this should go in `#if ANDROID` as checking whether the file exists isn't useful on the other platforms.

**Comment by oleavr** [comment]
Thanks a lot for doing this by the way! 🤘

**Comment by viniciusmarangoni** [comment]
I will try to fix the recommendations in this week. Thanks a lot.

**Comment by viniciusmarangoni** [comment]
Seems ok now. Could you please check it? Thanks.

File: lib/gadget/gadget.vala [hunk 9]
```diff
@@ -494,6 +494,18 @@ namespace Frida.Gadget {
 	private Config load_config (Location location) throws Error {
 		var config_path = derive_config_path_from_file_path (location.path);
 
+#if ANDROID
+		if (!FileUtils.test (config_path, FileTest.EXISTS)) {
+			var ext_index = config_path.last_index_of_char ('.');
+			if (ext_index != -1) {
+				config_path = config_path[0:ext_index] + ".config.so";
+			}
+			else{
```
Comment (oleavr) [inline lib/gadget/gadget.vala]: These should go on the same line, i.e.: `} else {`

**Comment by viniciusmarangoni** [comment]
Seems good now :)

**Comment by oleavr** [comment]
Thanks! 🤘
